testcase/testcase.go

```diff
@@ -254,6 +254,21 @@ func (tcs *TestCaseSet) Compare(events []logstash.Event, diffCommand []string, l
 		}
 	}()
 
+	// Sort the events and expected events so that they are both in the same order
+	// This is important as event ordering is not guaranteed under logstash config changes
+	sort.Slice(events, func(i, j int) bool {
+		iText, _ := json.Marshal(events[i])
+		jText, _ := json.Marshal(events[j])
+		return string(iText) < string(jText)
+	})
+
+	sortedEvents := tcs.ExpectedEvents
+	sort.Slice(sortedEvents, func(i, j int) bool {
+		iText, _ := json.Marshal(events[i])
+		jText, _ := json.Marshal(events[j])
+		return string(iText) < string(jText)
+	})
```

Comment:
There are a couple of problems with this code. First, slices are like pointers, so assigning one to the other doesn't make a copy of the original slice (this is unlike arrays). See https://play.golang.org/p/TcQbH3BtWw_Y. Instead you need to allocate a new slice of the same size as the original slice and use the copy() builtin.

    sortedEvents := make([]logstash.Event, len(tcs.ExpectedEvents))
    copy(sortedEvents, tcs.ExpectedEvents)

There's still a copy/paste mistake on lines 267-268; the `events` slice reference should be `sortedEvents`. Perhaps the test case could be adjusted to catch problems like this? Right now it apparently happens to work anyway.

Similarly you also need to copy `events` so we don't rearrange the elements of the input parameter. How about adding a function that takes a `[]logstash.Event` value and returns a sorted copy of it?

Unrelated to these lines, but while I'm at it, you've changed the mode of this file so that it's executable.

Reply:
Apologies it's taken so long to make these changes!  You can probably tell that Go is not my usual language.

I am having trouble building the project now so I've not been able to verify the changes yet.